cmake/JuliaUtils.cmake

```diff
@@ -0,0 +1,88 @@
+#
+
+include(CMakeVarMacros)
+
+get_filename_component(jl_cmake_utils_dir
+  "${CMAKE_CURRENT_LIST_FILE}" PATH)
+
+set(jl_cmake_utils_dir "${jl_cmake_utils_dir}" CACHE "" INTERNAL FORCE)
+
+function(jl_std_fpath var path)
+  if(NOT IS_ABSOLUTE "${path}")
+    set(${var} "${path}" PARENT_SCOPE)
+    return()
+  endif()
+  file(RELATIVE_PATH src_path "${CMAKE_CURRENT_SOURCE_DIR}" "${path}")
+  file(RELATIVE_PATH bin_path "${CMAKE_CURRENT_BINARY_DIR}" "${path}")
+  string(LENGTH "${src_path}" src_len)
+  string(LENGTH "${bin_path}" bin_len)
+  if(src_len GREATER bin_len)
+    set(${var} "${bin_path}" PARENT_SCOPE)
+  else()
+    set(${var} "${src_path}" PARENT_SCOPE)
+  endif()
+endfunction()
+
+function(jl_fpath_to_target var file)
+  jl_std_fpath(std_fname "${file}")
+  string(REGEX REPLACE "[^-_.a-zA-Z0-9]" "_" std_fname "${std_fname}")
+  string(REGEX REPLACE "_+" "_" std_fname "${std_fname}")
+  cmake_utils_get_unique_name("${std_fname}" target)
+  set("${var}" "${target}" PARENT_SCOPE)
+endfunction()
+
+function(jl_rewrite_dep_list var deps)
+  set(new_deps)
+  foreach(dep ${deps})
+    if(TARGET "${dep}")
+      set(new_deps ${new_deps} "$<TARGET_FILE:${dep}>")
+    elseif(IS_ABSOLUTE "${dep}")
+      set(new_deps ${new_deps} ${dep})
+    else()
+      get_filename_component(dep "${dep}" ABSOLUTE)
+      set(new_deps ${new_deps} ${dep})
+    endif()
+  endforeach()
+  set(${var} "${new_deps}" PARENT_SCOPE)
+endfunction()
+
+function(jl_rewrite_output_list var outputs)
+  set(new_outputs)
+  foreach(output ${outputs})
+    if(IS_ABSOLUTE "${output}")
+      set(new_outputs ${new_outputs} ${output})
+    else()
+      set(new_outputs ${new_outputs} "${CMAKE_CURRENT_BINARY_DIR}/${output}")
+    endif()
+  endforeach()
+  set(${var} "${new_outputs}" PARENT_SCOPE)
+endfunction()
+
+function(jl_custom_target target outputs output_deps target_deps autodep_file)
+  jl_rewrite_dep_list(output_deps "${output_deps}")
+  jl_rewrite_output_list(outputs "${outputs}")
+  if((NOT "${autodep_file}" STREQUAL "") AND
+      (NOT IS_ABSOLUTE "${autodep_file}"))
+    set(autodep_file "${CMAKE_CURRENT_BINARY_DIR}/${autodep_file}")
+  endif()
+  add_custom_target("${target}" ALL
+    COMMAND "${CMAKE_COMMAND}"
+    "-DOUTPUTS=${outputs}" "-DOUTPUT_DEPS=${output_deps}"
+    "-DAUTODEP_FILE=${autodep_file}" "-DCOMMAND_SPEC=${ARGN}"
+    "-DCUR_BIN_DIR=${CMAKE_CURRENT_BINARY_DIR}"
+    "-DCUR_SRC_DIR=${CMAKE_CURRENT_SOURCE_DIR}"
+    -P "${jl_cmake_utils_dir}/jl_custom_target.cmake"
+    DEPENDS ${target_deps} VERBATIM)
+endfunction()
```

Comment:
> 10 layers of custom cmake functions, including some nonstandard global mechanism

The global mechanism is actually pretty standard and is used in the builtin `FindGettext.cmake` module. (See `_GETTEXT_GET_UNIQUE_TARGET_NAME`).

It is used to generate a unique target name from the file name and is actually not used anymore after I realized that I'm explicitely supplying the target name anyway. So the only used function in this file is actually only the `jl_custom_target` and the two helper functions above, which I split out to make the logic more clear.

The other piece to get this working is `jl_custom_target.cmake` below, which is basically a wrapper that run the command and check the generated file list. Almost half of the file is actually output and formatting (everything before definition of `run_target` as well as `message()`'s below) so it shouldn't be too bad.

> You're modifying the source to log files in the sysimg

The biggest change is actually to find generated source file in the binary directory (check for `image_file` and `JL_BUILD_BASE_PATH`). The logging is only a very minimum function on top of that (since I've already got the full path in one case anyway).

> `CMAKE_CONFIGURE_DEPENDS`

This is actually already possible on older cmake version with `configure_file(COPYONLY)` which is what I'm using now for `VERSION`

> `BYPRODUCTS`

This seems to be Ninja-only and I don't think we need it.